### PR TITLE
Test fitonecycle

### PR DESCRIPTION
**Pull request by Benudek**
PR for testing fit and fit_one_cycle with inspecting learning rate and momentum. This is a follow up PR of closed PR https://github.com/fastai/fastai/pull/1453/files , 

Test with 

- pytest tests/test_basic_train.py (add -s for screenoutput )
- pytest tests/test_train.py (add -s for screenoutput )

see forum thread here: https://forums.fast.ai/t/improving-expanding-functional-tests/32929/44

### CODE REVIEW

**Comment by Benudek** [comment]
see forum thread here: https://forums.fast.ai/t/improving-expanding-functional-tests/32929/44

File: tests/test_train.py [hunk 1]
```diff
@@ -1,3 +1,106 @@
...
+
+@pytest.fixture(scope="module")
+def learn():
+    ## fixture for same model in fit and fit_one_cycle and demonstrate difference in behavior
+    ## test_fit and test_fit_one_cycle assert correctly with decommented models, 
+    ## fake_learner defaulted for performance reasons
+    
+    learn = fake_learner(50,50)
+
+    #path = untar_data(URLs.MNIST_TINY)
```
Comment (sgugger) [inline tests/test_train.py]: This is a lot of commented code that we don't need.

File: tests/test_train.py [hunk 2]
```diff
@@ -1,3 +1,106 @@
...
+
+    ## Test confirms learning rate and momentum are stable, see difference to test_fit_one_cycle
+    
+    learning_rate = 3e-3
+    weight_decay = 1e-2
+    eps = 4
+    
+    with CaptureStdout() as cs:  learn.fit(epochs=eps, lr=learning_rate, wd=weight_decay)
+    #assert_screenout(cs.out, str(eps))
+    lrs = get_learning_rates(learn)
```
Comment (sgugger) [inline tests/test_train.py]: This is awfully long to just test if learn.recorder.lrs only has one element: `set(learn.recorder.lrs)` would be quicker, don't you think?

File: tests/test_train.py [hunk 3]
```diff
@@ -1,3 +1,106 @@
...
+    ## see graphical representation here: output cell 17 of, learn.sched.plot_lr() in 
+    ## https://github.com/sgugger/Deep-Learning/blob/master/Cyclical%20LR%20and%20momentums.ipynb
+
+    cycle_length = 4
+    lr = 3e-3 # 5e-3
+
+    with CaptureStdout() as cs: learn.fit_one_cycle(cycle_length, lr)
+    #assert_screenout(cs.out, str(cycle_length)) 
+
+    listlrs =  get_learning_rates(learn) #give_lrs(learn)
```
Comment (sgugger) [inline tests/test_train.py]: Why test only the lrs at the end of the epochs?

File: tests/utils/param.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+""" Helper functions for inspecting hyperparameters """
+
+from fastai.basics import *
+
+def get_lrsmoms_paired(learn):
```
Comment (sgugger) [inline tests/utils/param.py]: Not sure those changes are necessary: again why only test the lrs/moms at the end of each epoch only?

**Comment by sgugger** [comment]
Thanks! There are a few things I think we can make a little bit better here, added some comments.

File: tests/test_train.py (hunk 2 above)
Comment (Benudek) [reply tests/test_train.py]: Done (with list(...)). Had grabbed the pattern from test_callbacks_csv_logger.get_train_losses. But its s different context, so indeed should be shorter

File: tests/test_train.py (hunk 1 above)
Comment (Benudek) [reply tests/test_train.py]: eliminated

**Comment by Benudek** [comment]
adjusted

File: tests/test_train.py [hunk 5]
```diff
@@ -1,3 +1,83 @@
...
+
+    ## Test confirms learning rate and momentum are stable, see difference to test_fit_one_cycle
+    
+    learning_rate = 3e-3
+    weight_decay = 1e-2
+    eps = 4
+    
+    with CaptureStdout() as cs:  learn.fit(epochs=eps, lr=learning_rate, wd=weight_decay)
+   
+    lrs = list(learn.recorder.lrs)
```
Comment (sgugger) [inline tests/test_train.py]: I think you misunderstood my previous comment, we can test that we always used the same lr by checking `set(learn.recorder.lrs) == {learning_rate}`.  Same for the momentum!

File: tests/test_train.py [hunk 6]
```diff
@@ -1,3 +1,83 @@
...
+
+    cycle_length = 4
+    lr = 3e-3
+
+    with CaptureStdout() as cs: learn.fit_one_cycle(cycle_length, lr)
+  
+    listlrs = list(learn.recorder.lrs)
+    listmoms = list(learn.recorder.moms) 
+    
+    ## eliminate the final low learning rates with corresponding momentum
```
Comment (sgugger) [inline tests/test_train.py]: Why do you do this? It's normal we get fom init_lr to max_lr then 0 (not exactly 0 as we stop just before but almost).

File: tests/test_train.py [hunk 7]
```diff
@@ -1,3 +1,83 @@
 import pytest,fastai
+from utils.fakes import *
+from utils.text import *
 
-#I'm all empty, please fill me with tests about the functions in train.py
\ No newline at end of file
+## filename: test_train.py
+## tests functions in train.py
```
Comment (sgugger) [inline tests/test_train.py]: Last note: please remove all useless blank lines to save vertical space (fastai style guide), thanks!

File: tests/test_train.py (hunk 5 above)
Comment (Benudek) [reply tests/test_train.py]: ah, okay. Yes - understood now and adjusted (less code indeed)

**Comment by Benudek** [comment]
is this fine now Sylvain? ;-)

**Comment by sgugger** [comment]
Looking good now, thanks!

**Comment by stas00** [comment]
> great, btw. If there is sth around PR etiquette to do better, I am all ears: https://forums.fast.ai/t/developer-chat/22363/715?u=benudek

https://forums.fast.ai/t/developer-chat/22363/720

> Overall, I hope this is useful (and not causing too much effort to review) for the library, for me it's a useful learning curve ;-)

We have all been sorting it out, it's a normal process when a library goes from 1-2 developers to many. So over time I'm sure better guidelines will be created. Until then watching the common style and using common sense will lead you far.
